File: src/usecase/runTrendBatch.ts
import {
  AppState,
  DEFAULT_MAX_TOPICS,
  EnrichedCluster,
  MAX_TOPICS_LIMIT
} from "../domain/model.js";
import { clusterByUrl, scoreItems } from "../domain/ranking.js";
import { renderMarkdown } from "../domain/markdown.js";
import { AIProvider } from "../ports/aiProvider.js";
import { Notifier } from "../ports/notifier.js";
import { SourceFetcher } from "../ports/sourceFetcher.js";
import { StateStore } from "../ports/stateStore.js";
import { sha256 } from "../shared/hash.js";
import { Logger } from "../shared/logger.js";
import { shouldPostOnDate, toJstDateString } from "../shared/time.js";

export interface RunOptions {
  mode: "normal" | "force";
  dryRun: boolean;
  maxTopics: number;
  debug: boolean;
  runId: string;
  audienceProfile: string;
  now?: Date;
}

interface Dependencies {
  fetchers: SourceFetcher[];
  stateStore: StateStore;
  aiProvider: AIProvider;
  notifier: Notifier;
  logger: Logger;
}

function clampTopics(maxTopics: number): number {
  if (!Number.isFinite(maxTopics)) return DEFAULT_MAX_TOPICS;
  return Math.max(1, Math.min(MAX_TOPICS_LIMIT, Math.floor(maxTopics)));
}

export async function runTrendBatch(deps: Dependencies, options: RunOptions): Promise<void> {
  const now = options.now ?? new Date();
  const jstDate = toJstDateString(now);
  const shouldPost = options.mode === "force" ? true : shouldPostOnDate(jstDate);

  const state = await deps.stateStore.load();
  const logBase = {
    run_id: options.runId,
    mode: options.mode,
    jst_date: jstDate,
    schedule_match: shouldPost
  };
  deps.logger.info("run start", logBase);

  if (options.mode === "normal" && !shouldPost) {
    deps.logger.info("skip by alternate-day rule", logBase);
    return;
  }

  if (state.postedHashes[jstDate] && options.mode === "normal") {
    deps.logger.info("skip by duplicate guard", logBase);
    return;
  }

  const fetchResults = await Promise.allSettled(
    deps.fetchers.map((fetcher) => fetcher.fetchItems(now.toISOString()))
  );
  const sourceCounts: Record<string, number> = {};
  const sourceItems = [];
  for (let i = 0; i < fetchResults.length; i += 1) {
    const fetcher = deps.fetchers[i];
    const res = fetchResults[i];
    if (!fetcher) continue;
    if (res?.status === "fulfilled") {
      sourceCounts[fetcher.name] = res.value.length;
      sourceItems.push(...res.value);
    } else {
      sourceCounts[fetcher.name] = 0;
      deps.logger.error("source failed", { source: fetcher.name, error: String(res?.reason) });
    }
  }

  const scored = scoreItems(sourceItems, now);
  const clusters = clusterByUrl(scored);
  const maxTopics = clampTopics(options.maxTopics);

  let enriched: EnrichedCluster[] = [];
  let aiFallback = false;
  try {
    enriched = await deps.aiProvider.enrich(clusters, options.audienceProfile);
  } catch (err) {
    aiFallback = true;
    deps.logger.error("ai enrich failed; fallback enabled", { error: String(err) });
  }

  const importantClusterIds = new Set(enriched.filter((e) => e.isImportant).map((e) => e.clusterId));
  let outputClusters = clusters.filter((c) => importantClusterIds.has(c.id)).slice(0, maxTopics);
  let outputEnriched = enriched.filter((e) => importantClusterIds.has(e.clusterId));

  if (outputClusters.length === 0) {
    outputClusters = clusters.slice(0, maxTopics);
    const existing = new Map(enriched.map((e) => [e.clusterId, e]));
    outputEnriched = outputClusters.map((c) => {
      const found = existing.get(c.id);
      if (found) return { ...found, isImportant: true };
      return {
        clusterId: c.id,
        summaryJa: "要約生成に失敗したため、元リンクを確認してください。",
        tags: [],
        reasonToRead: "対象読者の実務に関連する可能性が高いため。",
        isImportant: true
      };
    });
  }

  const markdown = renderMarkdown({
    jstDate,
    audienceProfile: options.audienceProfile,
    clusters: outputClusters,
    enriched: outputEnriched
  });

  const postHash = sha256(markdown);
  if (state.postedHashes[jstDate] === postHash && options.mode === "normal") {
    deps.logger.info("skip by same hash guard", { ...logBase, hash: postHash });
    return;
  }

  if (options.dryRun) {
    deps.logger.info("dryRun markdown", { ...logBase, markdown });
  } else {
    const result = await deps.notifier.notify(markdown);
    deps.logger.info("slack result", { ...logBase, status: result.status, ok: result.ok });
    if (!result.ok) {
      throw new Error(`slack webhook failed: ${result.status} ${result.body}`);
    }
  }

  const nextState: AppState = {
    lastRunJstDate: jstDate,
    itemsState: {
      ...state.itemsState,
      ...Object.fromEntries(
        scored.map((x) => [
          `${x.source}:${x.id}`,
          {
            score: x.score,
            comments: x.comments,
            lastSeen: now.toISOString()
          }
        ])
      )
    },
    postedHashes: {
      ...state.postedHashes,
      [jstDate]: postHash
    }
  };

  await deps.stateStore.save(nextState);
  deps.logger.info("run complete", {
    ...logBase,
    fetched: sourceCounts,
    clustered_count: clusters.length,
    important_count: importantClusterIds.size,
    selected_count: outputClusters.length,
    ai_calls: 1,
    ai_fallback: aiFallback
  });
}

File: internal/usecase/run_trend_batch.go
package usecase

import (
	"context"
	"errors"
	"time"

	"github.com/HiroVodka/trendog/internal/domain"
	"github.com/HiroVodka/trendog/internal/ports"
	"github.com/HiroVodka/trendog/internal/shared"
)

type RunOptions struct {
	Mode            string
	DryRun          bool
	MaxTopics       int
	Debug           bool
	RunID           string
	AudienceProfile string
	Now             time.Time
}

type Dependencies struct {
	Fetchers   []ports.SourceFetcher
	StateStore ports.StateStore
	AI         ports.AIProvider
	Notifier   ports.Notifier
	Logger     shared.Logger
}

func RunTrendBatch(ctx context.Context, deps Dependencies, opts RunOptions) error {
	now := opts.Now
	if now.IsZero() {
		now = time.Now().UTC()
	}
	jstDate := shared.ToJSTDateString(now)
	shouldPost := opts.Mode == "force" || shared.ShouldPostOnDate(jstDate)

	state, err := deps.StateStore.Load(ctx)
	if err != nil {
		return err
	}
	logBase := map[string]interface{}{"run_id": opts.RunID, "mode": opts.Mode, "jst_date": jstDate, "schedule_match": shouldPost}
	deps.Logger.Info("run start", logBase)

	if opts.Mode == "normal" && !shouldPost {
		deps.Logger.Info("skip by alternate-day rule", logBase)
		return nil
	}
	if _, ok := state.PostedHashes[jstDate]; ok && opts.Mode == "normal" {
		deps.Logger.Info("skip by duplicate guard", logBase)
		return nil
	}

	sourceCounts := map[string]int{}
	allItems := make([]domain.SourceItem, 0, 256)
	for _, f := range deps.Fetchers {
		items, ferr := f.FetchItems(ctx, now.Format(time.RFC3339))
		if ferr != nil {
			sourceCounts[f.Name()] = 0
			deps.Logger.Error("source failed", map[string]interface{}{"source": f.Name(), "error": ferr.Error()})
			continue
		}
		sourceCounts[f.Name()] = len(items)
		allItems = append(allItems, items...)
	}

	scored := domain.ScoreItems(allItems, now)
	clusters := domain.ClusterByURL(scored)
	maxTopics := clampTopics(opts.MaxTopics)

	enriched, aiFallback := enrichWithFallback(ctx, deps, clusters, opts.AudienceProfile)
	importantIDs := map[string]bool{}
	for _, e := range enriched {
		if e.IsImportant {
			importantIDs[e.ClusterID] = true
		}
	}

	selected := make([]domain.Cluster, 0, maxTopics)
	for _, c := range clusters {
		if importantIDs[c.ID] {
			selected = append(selected, c)
			if len(selected) >= maxTopics {
				break
			}
		}
	}
	if len(selected) == 0 {
		if len(clusters) > maxTopics {
			selected = clusters[:maxTopics]
		} else {
			selected = clusters
		}
	}

	selectedSet := map[string]bool{}
	for _, c := range selected {
		selectedSet[c.ID] = true
	}
	outEnriched := make([]domain.EnrichedCluster, 0, len(selected))
	for _, e := range enriched {
		if selectedSet[e.ClusterID] {
			outEnriched = append(outEnriched, e)
		}
	}
	for _, c := range selected {
		found := false
		for _, e := range outEnriched {
			if e.ClusterID == c.ID {
				found = true
				break
			}
		}
		if !found {
			outEnriched = append(outEnriched, domain.EnrichedCluster{ClusterID: c.ID, IsImportant: true, ReasonToRead: "対象読者の実務に関連する可能性が高いため。"})
		}
	}

	md := domain.RenderMarkdown(jstDate, opts.AudienceProfile, selected, outEnriched)
	hash := shared.SHA256(md)
	if state.PostedHashes[jstDate] == hash && opts.Mode == "normal" {
		deps.Logger.Info("skip by same hash guard", map[string]interface{}{"run_id": opts.RunID, "hash": hash})
		return nil
	}

	if opts.DryRun {
		deps.Logger.Info("dryRun markdown", map[string]interface{}{"run_id": opts.RunID, "markdown": md})
	} else {
		res, nerr := deps.Notifier.Notify(ctx, md)
		if nerr != nil {
			return nerr
		}
		deps.Logger.Info("slack result", map[string]interface{}{"run_id": opts.RunID, "status": res.Status, "ok": res.OK})
		if !res.OK {
			return errors.New("slack webhook failed")
		}
	}

	if state.ItemsState == nil {
		state.ItemsState = map[string]domain.ItemState{}
	}
	for _, s := range scored {
		k := string(s.Source) + ":" + s.ID
		state.ItemsState[k] = domain.ItemState{Score: s.Score, Comments: s.Comments, LastSeen: now.Format(time.RFC3339)}
	}
	if state.PostedHashes == nil {
		state.PostedHashes = map[string]string{}
	}
	state.LastRunJSTDate = jstDate
	state.PostedHashes[jstDate] = hash
	if err := deps.StateStore.Save(ctx, state); err != nil {
		return err
	}

	deps.Logger.Info("run complete", map[string]interface{}{
		"run_id": opts.RunID, "mode": opts.Mode, "jst_date": jstDate,
		"fetched": sourceCounts, "clustered_count": len(clusters), "selected_count": len(selected),
		"important_count": len(importantIDs), "ai_fallback": aiFallback,
	})
	return nil
}

func enrichWithFallback(ctx context.Context, deps Dependencies, clusters []domain.Cluster, audience string) ([]domain.EnrichedCluster, bool) {
	enriched, err := deps.AI.Enrich(ctx, clusters, audience)
	if err != nil {
		deps.Logger.Error("ai enrich failed; fallback enabled", map[string]interface{}{"error": err.Error()})
		return nil, true
	}
	return enriched, false
}

func clampTopics(v int) int {
	if v <= 0 {
		return domain.DefaultMaxTopics
	}
	if v > domain.MaxTopicsLimit {
		return domain.MaxTopicsLimit
	}
	return v
}
